src/index/TextIndexReadWrite.cpp

```diff
@@ -0,0 +1,185 @@
+    meta._startScorelist = currentOffset;
+    meta._lastByte = currentOffset - 1;
+    return meta;
+  }
+
+  auto firstElements =
+      postings | ql::views::transform([](const Posting& posting) {
+        return std::get<0>(posting).get();
+      });
+
+  auto secondElements =
+      postings | ql::views::transform([](const Posting& posting) {
+        return std::get<1>(posting);
+      });
+
+  auto thirdElements =
+      postings | ql::views::transform([](const Posting& posting) {
+        return std::get<2>(posting);
+      });
+
+  std::vector<uint64_t> textRecordList(firstElements.begin(),
+                                       firstElements.end());
+  std::vector<WordIndex> wordIndexList(secondElements.begin(),
+                                       secondElements.end());
+  std::vector<Score> scoreList(thirdElements.begin(), thirdElements.end());
+
+  GapEncode textRecordEncoder(textRecordList);
+  FrequencyEncode wordIndexEncoder(wordIndexList);
+  FrequencyEncode scoreEncoder(scoreList);
+
+  meta._startContextlist = currentOffset;
+  textRecordEncoder.writeToFile(out, meta._nofElements, currentOffset);
+
+  meta._startWordlist = currentOffset;
+  if (!skipWordlistIfAllTheSame || wordIndexEncoder.getCodeBook().size() > 1) {
+    wordIndexEncoder.writeToFile(out, meta._nofElements, currentOffset);
+  }
+
+  meta._startScorelist = currentOffset;
+  scoreEncoder.writeToFile(out, meta._nofElements, currentOffset);
+
+  meta._lastByte = currentOffset - 1;
+
+  return meta;
+}
+
+// ____________________________________________________________________________
+template <typename T>
+size_t writeCodebook(const vector<T>& codebook, ad_utility::File& file) {
+  size_t byteSizeOfCodebook = sizeof(T) * codebook.size();
+  file.write(&byteSizeOfCodebook, sizeof(byteSizeOfCodebook));
+  file.write(codebook.data(), byteSizeOfCodebook);
+  return byteSizeOfCodebook + sizeof(byteSizeOfCodebook);
+}
+
+// ____________________________________________________________________________
+template <typename Numeric>
+size_t writeList(const std::vector<Numeric> data, size_t nofElements,
+                 ad_utility::File& file) {
+  if (nofElements > 0) {
+    std::vector<uint64_t> encoded;
+    encoded.resize(nofElements);
+    size_t size = ad_utility::Simple8bCode::encode(data.data(), nofElements,
+                                                   encoded.data());
+    size_t ret = file.write(encoded.data(), size);
+    AD_CONTRACT_CHECK(size == ret);
+    return size;
+  } else {
+    return 0;
+  }
+}
+
+// ____________________________________________________________________________
+template <typename T>
+void writeVectorAndMoveOffset(const std::vector<T>& vectorToWrite,
+                              size_t nofElements, ad_utility::File& file,
+                              off_t& currentOffset) {
+  size_t bytes =
+      textIndexReadWrite::writeList(vectorToWrite, nofElements, file);
+  currentOffset += bytes;
+}
+
```

Comment:
In a separate PR, you can refactor this to use our serialization library, which makes the reading and writing of such types to and from disk much more readable.